### PR TITLE
Add sso sample for skillhost-skill scenario

### PR DESCRIPTION
**Pull request by swagatmishra2007**
Checking into experimental for now, until it can be polished to be in the main samples repo

### CODE REVIEW

**Comment by gabog** [comment]
Hi @stevengum, the PR in the SDK repo shouldn't block the sample if it is good to go, I will work with swagat on that one separately.

**Comment by swagatmishra2007** [comment]
@stevengum can you remove your blocking the PR, as Gabo clarified that this is unrelated to the SDK PR?

File: experimental/sso-with-skills/RootBot/AdapterWithErrorHandler.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
+using Microsoft.Bot.Builder.Skills;
+using Microsoft.Bot.Builder.TraceExtensions;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.BotBuilderSamples.SimpleRootBot.Bots;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.BotBuilderSamples.SimpleRootBot
```
Comment (carlosscastro) [inline experimental/sso-with-skills/RootBot/AdapterWithErrorHandler.cs:18]: update namespaces to RootBot

File: experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            }
+
+            return _skillsConfig.Skills.Values.FirstOrDefault(s => string.Equals(s.AppId, appId, StringComparison.InvariantCultureIgnoreCase));
+        }
+
+        private async Task<bool> InterceptOAuthCards(ClaimsIdentity claimsIdentity, Activity activity)
+        {
+            if (activity.Attachments != null)
+            {
+                foreach (var attachment in activity.Attachments.Where(a => a?.ContentType == OAuthCard.ContentType))
```
Comment (carlosscastro) [inline experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs:94]: Consider FirstOrDefault(a => a.ContentType...)

File: experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                {
+                    var targetSkill = GetCallingSkill(claimsIdentity);
+
+                    if (targetSkill != null)
+                    {
+                        var oauthCard = ((JObject)attachment.Content).ToObject<OAuthCard>();
+
+                        if (oauthCard.TokenExchangeResource != null /*&& _tokenExchangeConfig.ProviderId == oauthCard.TokenExchangeResource.ProviderId*/)
```
Comment (carlosscastro) [inline experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs:102]: remove commented code

File: experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+            var skillConversationReference = await _conversationIdFactory.GetSkillConversationReferenceAsync(incomingActivity.Conversation.Id, cancellationToken).ConfigureAwait(false);
+            activity.Conversation = skillConversationReference.ConversationReference.Conversation;
+            activity.ServiceUrl = skillConversationReference.ConversationReference.ServiceUrl;
+
+            // route the activity to the skill
+            var response = await _skillClient.PostActivityAsync(_botId, targetSkill, _skillsConfig.SkillHostEndpoint, activity, cancellationToken);
+
+            // Check response status: true if success, false if failure
+            return (response.Status >= 200 && response.Status <= 299);
```
Comment (carlosscastro) [inline experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs:158]: HttpStatusCodes.OK, etc

File: experimental/sso-with-skills/SkillBot/Bots/ChildBot.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            await _conversationState.LoadAsync(turnContext, true, cancellationToken);
+            await _userState.LoadAsync(turnContext, true, cancellationToken);
+            try
+            {
+                await _dialog.RunAsync(turnContext, _conversationState.CreateProperty<DialogState>(nameof(DialogState)), cancellationToken);
+            }
+            catch
+            {
+                // TODO: remove try/catch once the service is no longer throwing for 'consent required' during ExchangeTokenAsync.
```
Comment (carlosscastro) [inline experimental/sso-with-skills/SkillBot/Bots/ChildBot.cs:44]: Remove TODO and try/catch

File: experimental/sso-with-skills/SkillBot/Dialogs/SignInDialog.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
+﻿using System;
```
Comment (carlosscastro) [inline experimental/sso-with-skills/SkillBot/Dialogs/SignInDialog.cs:1]: copyright

File: experimental/sso-with-skills/SkillBot/Dialogs/MainDialog.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
```
Comment (carlosscastro) [inline experimental/sso-with-skills/SkillBot/Dialogs/MainDialog.cs:1]: copyright

File: experimental/sso-with-skills/SkillBot/Dialogs/DisplayTokenDialog.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
```
Comment (carlosscastro) [inline experimental/sso-with-skills/SkillBot/Dialogs/DisplayTokenDialog.cs:1]: copyright

File: experimental/sso-with-skills/SkillBot/Dialogs/SignOutDialog.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Threading;
```
Comment (carlosscastro) [inline experimental/sso-with-skills/SkillBot/Dialogs/SignOutDialog.cs:1]: copyright

File: experimental/sso-with-skills/SkillBot/SkillAdapterWithErrorHandler.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, $"Exception caught on attempting to Delete ConversationState : {ex}");
+                }
+            };
+        }
+
+        public override async Task<InvokeResponse> ProcessActivityAsync(ClaimsIdentity claimsIdentity, Activity activity, BotCallbackHandler callback, CancellationToken cancellationToken)
+        {
+            if (!claimsIdentity.Claims.Any(x => x.Type == "azp"))
```
Comment (carlosscastro) [inline experimental/sso-with-skills/SkillBot/SkillAdapterWithErrorHandler.cs:73]: We want to remove this workaround as this is now natively supported by the SDK and Token service

File: experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs (hunk 4 above)
Comment (swagatmishra2007) [reply experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs:158]: will use IsSuccessCode from InvokeResponse once R9 is released

**Pull request merged**